From 70795cc5955ee31939b49d3860fee6ee74d07b10 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Tue, 15 Mar 2022 11:52:35 +0200
Subject: [PATCH] [SW-78885] gaudi3: initialize arc1 iccm memory

ARC1 ICCM is used by ROM to download and validate boot
descriptor image and is also used to copy boot descriptor data
while using boot info. In order to prevent any uninitialized read access
to iccm in ROM code, this patch performs init of ARC1 iccm memory only in
cold reset using memcopy block.

Change-Id: I60218f3cd8a9ca1dd9275586ac573f5182a7374e
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index e4cf6f5..d11028b 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -5,6 +5,7 @@
  */
 
 #include <arcboot.h>
+#include <delay.h>
 #include <gaudi3_fw_if.h>
 #include <efuse.h>
 #include <dev_spi_flash.h>
@@ -20,6 +21,64 @@
 #include <d0_psoc_global_conf_masks.h>
 
 void arcboot_rom_bringup_arcboot_ram(void) {}
+
+int arcboot_rom_mem_init(uint64_t addr, uint32_t size)
+{
+	uint32_t retry = HL_INIT_MEM_RETRIES;
+	uint32_t mem_cfg, cpy_sts;
+
+	regmap_writel((uint32_t)addr, mmD0_PSOC_GLOBAL_CONF_MEM_CPY_DEST_ADDR_L);
+	regmap_writel((uint32_t)(addr >> 32), mmD0_PSOC_GLOBAL_CONF_MEM_CPY_DEST_ADDR_H);
+
+	mem_cfg = regmap_readl(mmD0_PSOC_GLOBAL_CONF_MEM_CPY_CTRL2);
+	mem_cfg &= ~(D0_PSOC_GLOBAL_CONF_MEM_CPY_CTRL2_MEM_SIZE_M);
+
+	/* set size of in granuality of 16 bytes */
+	size = size >> 4;
+	mem_cfg |= (size & D0_PSOC_GLOBAL_CONF_MEM_CPY_CTRL2_MEM_SIZE_M);
+	mem_cfg |= D0_PSOC_GLOBAL_CONF_MEM_CPY_CTRL2_USE_CONST_M;
+
+	regmap_writel(mem_cfg, mmD0_PSOC_GLOBAL_CONF_MEM_CPY_CTRL2);
+
+	/* Enable copy and wait for completion */
+	regmap_writel(D0_PSOC_GLOBAL_CONF_MEM_CPY_CTRL_EN_M, mmD0_PSOC_GLOBAL_CONF_MEM_CPY_CTRL);
+	while (1) {
+		cpy_sts = regmap_readl(mmD0_PSOC_GLOBAL_CONF_MEM_CPY_STATUS);
+		if (cpy_sts & D0_PSOC_GLOBAL_CONF_MEM_CPY_STATUS_DONE_M)
+			break;
+
+		retry--;
+		if (!retry)
+			return -1;
+
+		udelay(HL_INIT_MEM_UDELAY);
+	}
+
+	return 0;
+}
+
+static void arcboot_rom_iccm_init(void)
+{
+	struct gaudi3_cold_rst_data cold_rst_data;
+	int ret;
+
+	if (arcboot_rom_global.die_id)
+		return;
+
+	/* Perform arc1 iccm init only on cold reset */
+	cold_rst_data.data = regmap_readl(mmCOLD_RST_FLOPS_2);
+	if (cold_rst_data.arc1_iccm_init_done)
+		return;
+
+	ret = arcboot_rom_mem_init((HL_GEN_MEM_BASE_FULL + HL_ARC1_ICCM_OFFSET), HL_ICCM_SIZE);
+	if (ret) {
+		arcboot_rom_status_update(PROGRESS_INIT_ARC1_ICCM_INIT_FAILED);
+		return;
+	}
+
+	cold_rst_data.arc1_iccm_init_done = 1;
+	regmap_writel(cold_rst_data.data, mmCOLD_RST_FLOPS_2);
+}
 
 void arcboot_rom_clear_sw_status(void)
 {
@@ -121,6 +180,8 @@
 
 	arcboot_rom_read_cold_flops();
 
+	arcboot_rom_iccm_init();
+
 	spi_flash_init();
 
 	if (!boot_desc_init(HL_FW_BOOT_PTR_SPI))
diff --git a/include/arcboot.h b/include/arcboot.h
index ca074ce..0267a39 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -60,6 +60,8 @@
 
 #define HL_INIT_SP_SRAM_RETRIES			100
 #define HL_INIT_SP_SRAM_UDELAY			10
+#define HL_INIT_MEM_RETRIES			100
+#define HL_INIT_MEM_UDELAY			10
 #define SP_SRAM_SIZE_16B_GRANULARITY		0x1000
 #define CRC32_SIZE		4
 
@@ -92,67 +94,68 @@
 enum arcboot_rom_progress {
 	PROGRESS_INIT = 0,
 	PROGRESS_INIT_SP_SRAM_NOT_COMPLETED = 1,
-	PROGRESS_INIT_SPI_BOOT_DESC_DMA_START = 2,
-	PROGRESS_INIT_SPI_BOOT_DESC_FAIL = 3,
-	PROGRESS_INIT_SPI_BOOT_DESC_DMA_FINISH = 4,
-	PROGRESS_INIT_EFUSE_SPI = 5,
-	PROGRESS_REVOKE_LOC = 6,
-	PROGRESS_REVOKE_LOC_0 = 7,
-	PROGRESS_REVOKE_LOC_1 = 8,
-	PROGRESS_KEY_LOC = 9,
-	PROGRESS_KEY_LOC_0 = 10,
-	PROGRESS_KEY_LOC_1 = 11,
-	PROGRESS_VER_LOC = 12,
-	PROGRESS_VER_LOC_0 = 13,
-	PROGRESS_VER_LOC_1 = 14,
-	PROGRESS_BRINGUP_PCI_PRI = 15,
-	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 16,
-	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 17,
-	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 18,
-	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 19,
-	PROGRESS_BRINGUP_PCI_PRI_DMA = 20,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 21,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 22,
-	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 23,
-	PROGRESS_BRINGUP_PCI_SEC = 24,
-	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 25,
-	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 26,
-	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 27,
-	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 28,
-	PROGRESS_BRINGUP_PCI_SEC_DMA = 29,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 30,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 31,
-	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 32,
-	PROGRESS_PCIE_IN_RECOVERY = 33,
-	PROGRESS_PPBOOT_START = 34,
-	PROGRESS_PPBOOT_FINISH = 35,
-	PROGRESS_BRINGUP_RAM_PRI = 36,
-	PROGRESS_BRINGUP_RAM_PRI_DMA = 37,
-	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 38,
-	PROGRESS_BRINGUP_RAM_SEC = 39,
-	PROGRESS_BRINGUP_RAM_SEC_DMA = 40,
-	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 41,
-	PROGRESS_ARCBOOT_RAM_START = 42,
-	PROGRESS_ARCBOOT_RECOVERY = 43,
-	PROGRESS_ARCBOOT_RECOVERY_DMA = 44,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR = 45,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_START = 46,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_FINISH = 47,
-	PROGRESS_ARCBOOT_EROM_PCIE = 48,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_START = 49,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_FINISH = 50,
-	PROGRESS_ARCBOOT_EROM_PCIE_START = 51,
-	PROGRESS_ARCBOOT_EROM_PCIE_FINISH = 52,
-	PROGRESS_ARCBOOT_EROM_RECOVERY = 53,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_START = 54,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_FINISH = 55,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_START = 56,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_FINISH = 57,
-	PROGRESS_ARCBOOT_EROM_TPM = 58,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_START = 59,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_FINISH = 60,
-	PROGRESS_ARCBOOT_EROM_TPM_START = 61,
-	PROGRESS_ARCBOOT_EROM_TPM_FINISH = 62,
+	PROGRESS_INIT_ARC1_ICCM_INIT_FAILED = 2,
+	PROGRESS_INIT_SPI_BOOT_DESC_DMA_START = 3,
+	PROGRESS_INIT_SPI_BOOT_DESC_FAIL = 4,
+	PROGRESS_INIT_SPI_BOOT_DESC_DMA_FINISH = 5,
+	PROGRESS_INIT_EFUSE_SPI = 6,
+	PROGRESS_REVOKE_LOC = 7,
+	PROGRESS_REVOKE_LOC_0 = 8,
+	PROGRESS_REVOKE_LOC_1 = 9,
+	PROGRESS_KEY_LOC = 10,
+	PROGRESS_KEY_LOC_0 = 11,
+	PROGRESS_KEY_LOC_1 = 12,
+	PROGRESS_VER_LOC = 13,
+	PROGRESS_VER_LOC_0 = 14,
+	PROGRESS_VER_LOC_1 = 15,
+	PROGRESS_BRINGUP_PCI_PRI = 16,
+	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 17,
+	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 18,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 19,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 20,
+	PROGRESS_BRINGUP_PCI_PRI_DMA = 21,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 22,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 23,
+	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 24,
+	PROGRESS_BRINGUP_PCI_SEC = 25,
+	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 26,
+	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 27,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 28,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 29,
+	PROGRESS_BRINGUP_PCI_SEC_DMA = 30,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 31,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 32,
+	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 33,
+	PROGRESS_PCIE_IN_RECOVERY = 34,
+	PROGRESS_PPBOOT_START = 35,
+	PROGRESS_PPBOOT_FINISH = 36,
+	PROGRESS_BRINGUP_RAM_PRI = 37,
+	PROGRESS_BRINGUP_RAM_PRI_DMA = 38,
+	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 39,
+	PROGRESS_BRINGUP_RAM_SEC = 40,
+	PROGRESS_BRINGUP_RAM_SEC_DMA = 41,
+	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 42,
+	PROGRESS_ARCBOOT_RAM_START = 43,
+	PROGRESS_ARCBOOT_RECOVERY = 44,
+	PROGRESS_ARCBOOT_RECOVERY_DMA = 45,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR = 46,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_START = 47,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_FINISH = 48,
+	PROGRESS_ARCBOOT_EROM_PCIE = 49,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_START = 50,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_FINISH = 51,
+	PROGRESS_ARCBOOT_EROM_PCIE_START = 52,
+	PROGRESS_ARCBOOT_EROM_PCIE_FINISH = 53,
+	PROGRESS_ARCBOOT_EROM_RECOVERY = 54,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_START = 55,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_FINISH = 56,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_START = 57,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_FINISH = 58,
+	PROGRESS_ARCBOOT_EROM_TPM = 59,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_START = 60,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_FINISH = 61,
+	PROGRESS_ARCBOOT_EROM_TPM_START = 62,
+	PROGRESS_ARCBOOT_EROM_TPM_FINISH = 63,
 };
 
 enum security_location {
@@ -171,6 +174,7 @@
 void arcboot_rom_validation_fail(void);
 void arcboot_rom_clear_recovery_flag(void);
 void arcboot_rom_load_recovery(void);
+int arcboot_rom_mem_init(uint64_t addr, uint32_t size);
 
 #define CLEAR_DMA_SECURITY_PROGRESS 0x3
 
